### PR TITLE
feat: added scaleway_storage_object_bucket resource

### PR DESCRIPTION
**Pull request by DenBeke**

### CODE REVIEW

**Comment by DenBeke** [comment]
cfr. #153

File: scaleway/resource_storage_object_bucket.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The name of the bucket",
+			},
+			"acl": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "private",
+				Description: "ACL of the bucket: either 'public-read' or 'private'. Private by default.",
```
Comment (QuentinBrosse) [inline scaleway/resource_storage_object_bucket.go]: Is it a good practice to document the default value in the description? I can't find this practice in other providers...

> Description is used as the description for docs or asking for user input. It should be relatively short (a few sentences max) and should be formatted to fit a CLI.

File: scaleway/resource_storage_object_bucket.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	bucketName := d.Get("name").(string)
+
+	s3Client := m.(*Meta).s3Client
+
+	_, err := s3Client.ListObjects(&s3.ListObjectsInput{
+		Bucket: aws.String(bucketName),
+	})
+	if err != nil {
+		if serr, ok := err.(awserr.Error); ok && serr.Code() == s3.ErrCodeNoSuchBucket {
+			log.Printf("[ERROR] Bucket %q was not found - removing from state!", d.Get("name").(string))
```
Comment (QuentinBrosse) [inline scaleway/resource_storage_object_bucket.go]: The string cast is useless.

Comment (QuentinBrosse) [reply scaleway/resource_storage_object_bucket.go]: Please use [this logger](https://github.com/terraform-providers/terraform-provider-scaleway/blob/master/scaleway/logger.go#L17).

File: scaleway/resource_storage_object_bucket.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+		bucketName := d.Get("name").(string)
+		acl := d.Get("acl").(string)
+		s3Client := m.(*Meta).s3Client
+
+		_, err := s3Client.PutBucketAcl(&s3.PutBucketAclInput{
+			Bucket: aws.String(bucketName),
+			ACL:    aws.String(acl),
+		})
+		if err != nil {
+			log.Printf("[ERROR] Couldn't update bucket ACL: %s", err)
```
Comment (QuentinBrosse) [inline scaleway/resource_storage_object_bucket.go]: Please use [this logger](https://github.com/terraform-providers/terraform-provider-scaleway/blob/master/scaleway/logger.go#L17).

File: scaleway/resource_storage_object_bucket_test.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+		return fmt.Errorf("error getting client: %s", err)
+	}
+
+	listBucketResponse, err := s3client.ListBuckets(&s3.ListBucketsInput{})
+	if err != nil {
+		return fmt.Errorf("couldn't list buckets: %s", err)
+	}
+
+	for _, bucket := range listBucketResponse.Buckets {
+		log.Println(*bucket.Name)
```
Comment (QuentinBrosse) [inline scaleway/resource_storage_object_bucket_test.go]: Please use [this logger](https://github.com/terraform-providers/terraform-provider-scaleway/blob/master/scaleway/logger.go#L17).

Comment (QuentinBrosse) [reply scaleway/resource_storage_object_bucket_test.go]: What do you think about:
```go
log.Printf("Deleting %s bucket", *bucket.Name)
```
?

File: scaleway/config.go [hunk 5]
```diff
@@ -170,6 +183,52 @@ func (c *Config) GetDeprecatedClient() (*sdk.API, error) {
...
+	if err != nil {
+		return nil, err
+	}
+	s3client := s3.New(s)
+
+	return s3client, nil
+}
+
+// getS3Region returns the correct S3 region for object storage based on the current region
+func (c *Config) getS3Region() string {
```
[the remark this answers is not in the document]
Comment (DenBeke) [reply scaleway/config.go]: It is used elsewhere as well. Wasn't sure if I wanted to keep it.

**Comment by QuentinBrosse** [comment]
Remove WIP?

**Comment by DenBeke** [comment]
> Remove WIP?

Yeps, I'll update the PR with your feedback and then it's "ready" for me.

File: scaleway/resource_storage_object_bucket_test.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+				Config: testAccCheckScalewayStorageObjectBucket,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("scaleway_storage_object_bucket.base", "name", testBucketName),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckScalewayStorageObjectBucketDestroy(s *terraform.State) error {
```
Comment (QuentinBrosse) [inline scaleway/resource_storage_object_bucket_test.go]: Add a `testAccCheckScalewayStorageObjectBucketUpdate`

File: scaleway/config.go [hunk 7]
```diff
@@ -58,6 +65,12 @@ func (c *Config) Meta() (*Meta, error) {
 	}
 	meta.deprecatedClient = deprecatedClient
 
+	s3Client, err := c.GetS3Client()
+	if err != nil {
+		return nil, fmt.Errorf("error: cannot create S3 client: %s", err)
```
Comment (kindermoumoute) [inline scaleway/config.go]: This is already an error, remove `error: `

File: scaleway/config.go [hunk 8]
```diff
@@ -191,3 +250,41 @@ func readDeprecatedScalewayConfig(path string) (string, string, error) {
...
+	}
+
+	type resBody struct {
+		Token token `json:"token"`
+	}
+
+	url := "https://account.scaleway.com/tokens/" + c.SecretKey
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		panic(err)
```
Comment (kindermoumoute) [inline scaleway/config.go]: return err?

File: scaleway/config.go [hunk 9]
```diff
@@ -191,3 +250,41 @@ func readDeprecatedScalewayConfig(path string) (string, string, error) {
...
+	}
+
+	res, err := scwClient.Client.Do(req)
+	if err != nil {
+		panic(err)
```
Comment (kindermoumoute) [inline scaleway/config.go]: return err?

File: scaleway/config.go [hunk 10]
```diff
@@ -191,3 +250,41 @@ func readDeprecatedScalewayConfig(path string) (string, string, error) {
...
+	}
+
+	body, err := ioutil.ReadAll(res.Body)
+	defer res.Body.Close()
+	if err != nil {
+		panic(err)
```
Comment (kindermoumoute) [inline scaleway/config.go]: return err?

File: scaleway/config.go [hunk 11]
```diff
@@ -191,3 +250,41 @@ func readDeprecatedScalewayConfig(path string) (string, string, error) {
...
+	}
+
+	fmt.Println(string(body))
+
+	content := &resBody{}
+	err = json.Unmarshal(body, content)
+	if err != nil {
+		panic(err)
```
Comment (kindermoumoute) [inline scaleway/config.go]: return err?

File: scaleway/resource_storage_object_bucket.go (hunk 1 above)
Comment (DenBeke) [reply scaleway/resource_storage_object_bucket.go]: Probably not :p

Comment (kindermoumoute) [reply scaleway/resource_storage_object_bucket.go]: IIRC the default value is shown in the log and in the terraform schema while editing the file. 🤔

File: scaleway/resource_storage_object_bucket_test.go (hunk 4 above)
Comment (DenBeke) [reply scaleway/resource_storage_object_bucket_test.go]: 👍

File: scaleway/config.go [hunk 12]
```diff
@@ -164,3 +209,77 @@ func readDeprecatedScalewayConfig(path string) (string, string, error) {
 	}
 	return data.Token, data.Organization, nil
 }
+
+// getAccessKeyFromSecretKey returns the access key that is coupled to the current token/secret key in the client.
+func (m *Meta) getAccessKeyFromSecretKey(scwClient *sdk.API) (string, error) {
```
Comment (jerome-quere) [inline scaleway/config.go]: no need as this is a new ressource (no backward compatibility)

File: scaleway/config.go [hunk 13]
```diff
@@ -143,6 +170,24 @@ func (m *Meta) bootstrapDeprecatedClient() error {
 	return nil
 }
 
+// bootstrapS3Client initializes a new s3 client from the configuration.
+func (m *Meta) bootstrapS3Client() error {
+
+	client, err := m.createS3ClientForRegion(m.DefaultRegion)
```
Comment (jerome-quere) [inline scaleway/config.go]: Merge createS3ClientForRegion into this method

File: scaleway/config.go [hunk 14]
```diff
@@ -164,3 +209,77 @@ func readDeprecatedScalewayConfig(path string) (string, string, error) {
...
+	}
+
+	return content.Token.AccessKey, nil
+}
+
+// s3AccessKey contains the access key that is needed for S3.
+// This is a global variable so we only have to do one request to fetch the token.
+//
+// This will be removed in v2.
+var s3AccessKey string
```
Comment (jerome-quere) [inline scaleway/config.go]: Can be removed as

File: scaleway/helpers.go [hunk 15]
```diff
@@ -310,6 +310,24 @@ func getRegion(d terraformResourceData, meta *Meta) (utils.Region, error) {
 	return utils.Region(""), ErrRegionNotFound
 }
 
+// getRegion will try to guess the zone from the following:
+//  - zone field of the resource data
+//  - default zone from config
+func getRegion(d terraformResourceData, meta *Meta) (utils.Region, error) {
```
Comment (jerome-quere) [inline scaleway/helpers.go]: duplicate

File: scaleway/resource_storage_object_bucket.go [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The name of the bucket",
+			},
+			"acl": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "private",
+				Description: "ACL of the bucket: either 'public-read' or 'private'.",
```
Comment (jerome-quere) [inline scaleway/resource_storage_object_bucket.go]: Use a validation function `validations.StringInSlice`

File: scaleway/resource_storage_object_bucket.go [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+
+	_, err = s3Client.CreateBucket(&s3.CreateBucketInput{
+		Bucket: aws.String(bucketName),
+		ACL:    aws.String(acl),
+	})
+	if err != nil {
+		return err
+	}
+
+	d.SetId(newRegionalId(utils.Region(region), bucketName))
```
Comment (jerome-quere) [inline scaleway/resource_storage_object_bucket.go]: utils.Region(region) useless cast

File: scaleway/resource_storage_object_bucket.go [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+
+	return resourceScalewayStorageObjectBucketRead(d, m)
+}
+
+func resourceScalewayStorageObjectBucketRead(d *schema.ResourceData, m interface{}) error {
+	s3Client, _, bucketName, err := getS3ClientWithRegionAndID(d, m)
```
Comment (jerome-quere) [inline scaleway/resource_storage_object_bucket.go]: please use same interface os `getInstanceapiWithRegionAndID` (take and id not d)
Usefull for tests

File: scaleway/resource_storage_object_bucket_test.go [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+	testBucketName       = fmt.Sprintf("terraform-test-%d", time.Now().Unix())
+	testBucketNameAms    = testBucketName + "ams"
+	testBucketNamePar    = testBucketName + "par"
+	testBucketACL        = "private"
+	testBucketUpdatedACL = "public-read"
+)
+
+// Test configs
+var testAccCheckScalewayStorageObjectBucket = fmt.Sprintf(`
+resource "scaleway_storage_object_bucket" "base" {
```
Comment (jerome-quere) [inline scaleway/resource_storage_object_bucket_test.go]: Please indent

File: scaleway/resource_storage_object_bucket_test.go [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+resource "scaleway_storage_object_bucket" "use-region-par" {
+  name = "%s"
+  region = "nl-ams"
+}
+`, testBucketName, testBucketNameAms, testBucketNamePar)
+
+var testAccCheckScalewayStorageObjectBucketUpdate = fmt.Sprintf(`
+resource "scaleway_storage_object_bucket" "base" {
```
Comment (jerome-quere) [inline scaleway/resource_storage_object_bucket_test.go]: Please update

File: scaleway/resource_storage_object_bucket_test.go [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+)
+
+// Test configs
+var testAccCheckScalewayStorageObjectBucket = fmt.Sprintf(`
+resource "scaleway_storage_object_bucket" "base" {
+  name = "%s"
+  region = "nl-ams"
+}
+
+resource "scaleway_storage_object_bucket" "use-region-ams" {
```
Comment (jerome-quere) [inline scaleway/resource_storage_object_bucket_test.go]: `ams-bucket`

File: scaleway/resource_storage_object_helpers.go [hunk 22]
```diff
@@ -0,0 +1,48 @@
+package scaleway
```
Comment (jerome-quere) [inline scaleway/resource_storage_object_helpers.go]: resource_storage_object_helpers.go => helper_storage_object.go

File: scaleway/resource_storage_object_bucket_test.go (hunk 20 above)
Comment (jerome-quere) [reply scaleway/resource_storage_object_bucket_test.go]: please indent

File: scaleway/resource_storage_object_bucket.go [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+				ForceNew:    true,
+				Description: "The name of the bucket",
+			},
+			"acl": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "private",
+				Description: "ACL of the bucket: either 'public-read' or 'private'.",
+				ValidateFunc: validation.StringInSlice([]string{
+					"private",
```
Comment (jerome-quere) [inline scaleway/resource_storage_object_bucket.go]: I would only expose "public"  and "private" ?

Comment (jerome-quere) [reply scaleway/resource_storage_object_bucket.go]: let's merge we will see what storage team respond.

**Comment by gmile** [comment]
I am trying to import existing bucket into my tf state. How do I know the UUID of bucket?

**Comment by jerome-quere** [comment]
Hi @gmile

This is an issue on our side. Import should be done by bucket name.
I open #321 to trak progress, this fix should be included in the next release.

**Pull request merged**